File: Scala/src/main/scala/observer/StatisticsDisplay.scala
package observer

class StatisticsDisplay(weatherData: WeatherData) extends DisplayElement with Observer {
  weatherData.registerObserver(this)
  private var maxTemp: Float = 0.0f
  private var minTemp: Float = 200f
  private var tempSum: Float = 0.0f
  private var numReadings: Int = 0

  override def display(): Unit = println(s"평균/최고/최저 온도: ${tempSum/numReadings}/${maxTemp}/${minTemp}")

  override def update(): Unit = {
    val t = weatherData.getTemperature()
    tempSum += t
    numReadings += 1

    maxTemp = maxTemp.max(t)
    minTemp = minTemp.min(t)

    display()
  }
}

File: Java/Observer/src/StatisticsDisplay.java
public class StatisticsDisplay implements Observer, DisplayElement{
  private float maxTemp = 0.0f;
  private float minTemp = 200;
  private float tempSum= 0.0f;
  private int numReadings = 0;
  private WeatherData weatherData;

  public StatisticsDisplay(WeatherData weatherData) {
    this.weatherData = weatherData;
  }

  @Override
  public void update() {
    maxTemp = Math.max(maxTemp, getTemperature());
    minTemp = Math.min(minTemp, getTemperature());
    tempSum += getTemperature();
    numReadings++;

    display();
  }

  @Override
  public void display() {
    System.out.println("통계: 평균 온도 " + getAverageTemperature() + "F, 최고 온도 " + maxTemp + "F, 최저 온도 " + minTemp);
  }

  private float getTemperature() {
    return weatherData.getTemperature();
  }

  private float getAverageTemperature() {
    return tempSum / numReadings;
  }

}
